### ISSUE TITLE
[BUG] RMM_FREE of an invalid address is returning RMM_SUCCESS

### ISSUE THREAD

**Issue by abellina** [rapidsai/rmm#147]
If I try to rmm free an invalid address (note `c_stream` is 0):
`err = RMM_FREE(reinterpret_cast<void*>(100), c_stream);`

It prints this warning:
`warning: Cuda API error detected: cudaFree returned (0x11)`

but `err` is `RMM_SUCCESS`. I expected: `RMM_ERROR_CUDA_ERROR`.

This is a recent issue in branch-0.10, possibly related to https://github.com/rapidsai/rmm/pull/127.

This is some repro code, where we can't go into the if statement.

```
  cudaStream_t c_stream = reinterpret_cast<cudaStream_t>(0);
  rmmError_t err = RMM_FREE(reinterpret_cast<void*>(100), c_stream);
  if (err != RMM_SUCCESS) {
    std::cout <<"not successful free of invalid address" << err<<std::endl;
  }
```

**Comment by abellina** [comment]
I am going to close this, as this is the desired behavior (fyi @harrism)

**Comment by harrism** [comment]
My comment on the other thread was that it is not desired behavior for free to throw an exception. But it should not successfully free an invalid pointer. 

I think 0x11 = 17 is cudaErrorInvalidDevicePointer, which is expected for CUDA 10.0 and earlier (CUDA 10.1 returns cudaErrorInvalidValue).

I can't even find the string "warning" anywhere in the RMM source code!

**Comment by abellina** [comment]
I ran this code in cuda-gdb and that message is coming from there, when it detects a cuda error: https://github.com/NVIDIA/cuda-gdb/blob/cuda-toolkit-7.0/gdb/breakpoint.c#L5728

The fact that `rmm::free` returns a status code at the moment means it probably shouldn't return success in this case if it can be detected, and leave it up to the caller whether to throw or not. Though I am not sure if we are trying to move away from `rmm::free` returning anything at all, like the C library function `free`.

The odd thing as far as I can tell from some of the tests I ran is after this failure happens, other subsequent cuda calls are dying with "invalid device pointer". I haven't figured out why yet.. I'll try to come up with a c++ repro case.

**Comment by abellina** [comment]
Ok yeah I believe RMM should at least call `cudaGetLastError` to clear that as well, subsequent calls succeed after I call that manually after a free that would set the error.

**Comment by harrism** [comment]
Thanks for reopening. I found the root cause of the bug and will fix.

**Comment by harrism** [comment]
Turns out this can't be completely fixed until #38 is added. We switched to `device_memory_resource` internally, which is based on `std::pmr::memory_resource`. The spec for the latter specifies that `deallocate` should not throw, and is of `void` return type -- so no way to detect errors on deallocation (just like C `free()`).

For now, rmmFree returns CUDA_SUCCESS in pool mode even if the free fails. When #38 is merged, we will probably get rid of the return codes from RMM completely. I have added changes to #146 that cause rmmFree to return errors in non-pool mode, to partially fix this issue.

**Comment by jrhemstad** [comment]
> When #38 is merged, we will probably get rid of the return codes from RMM completely.

Agreed. I think this will be a "wontfix" issue. 

If deallocation fails there's not much we can do about it. Best I can think of is to print an error message to `stderr` in Debug mode.
